### PR TITLE
Add Scala Analytics SDK 0.4.0 post

### PR DESCRIPTION
**Pull request by chuwy**
Please assign me to review once ready, @rzats

### CODE REVIEW

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+4. [Upgrading](#upgrading)
+5. [Getting help](#help)
+
+<!--more-->
+
+<h2 id="event-api">1. Event API</h2>
+
+Previously, the JSON Event Transformer - an event transformer that takes a Snowplow enriched event and converts it into a JSON ready for further processing - used to return Strings, which represented enriched events turned into JSON objects. While this was a very unopinionated and minimalistic approach, it involved a lot of extra post-processing, namely:
+
+- While it was always known that the result was a JSON object, accessing individual fields required casting it to an instance of the json4s AST class via `parse(result)` or even `parse(result).asInstanceOf[JObject]`. Both of these operations were unsafe and could lead to runtime exceptions if not accessed correctly.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: I'd make these points slightly shorter. I totally agree with them, but there's shorter way to state that UX was suboptimal.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+- Several commonly accessed fields in enriched events were, by definition, always present in a valid row. However, accessing them still required redundant error processing logic that would be used for any other fields, e.g. `parsedJson.map("event_id").getOrElse(throw new RuntimeException("event_id is not present in the enriched event")`
+- Getting a list of shredded types required having to use a separate function, `jsonifyWithInventory`.
+
+In 0.4.0, the `EventTransformer` API has been fully deprecated in favor of `Event` - a single typesafe container that represents a canonical Snowplow event. This case class has primary functions:
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: > canonical Snowplow event.

Link to https://github.com/snowplow/snowplow/wiki/canonical-event-model

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+- `Event.parse(line)` - similar to the old `transform` function, it accepts an enriched Snowplow event as a string and returns an `Event` instance as a result, with fields that can be accessed similar to a regular Scala case class.
+- `event.toJson(lossy)` - similar to the old `getValidatedJsonEvent` function, it transforms an `Event` into a validated JSON whose keys are the field names corresponding to the EnrichedEvent POJO of the Scala Common Enrich project. If the lossy argument is true, any self-describing events in the fields (unstruct_event, contexts and derived_contexts) are returned in a "shredded" format, e.g. `"unstruct_event_com_acme_1_myField": "value"`. If it is set to false, they are not flattened into underscore-separated top-level fields, using a standard self-describing format instead.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: > similar

Must be identical.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+- `event.inventory()` - extracts metadata from the event containing information about the types and Iglu URIs of its shred properties (unstruct_event, contexts and derived_contexts). Unlike version 0.3.0, it no longer requires a `transformWithInventory` call and can be obtained from any `Event` instance.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: `()` at the end of method means it is side-effecting (same below)

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md (hunk 2 above)
Comment (chuwy) [reply _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: Also worth to mention that it has 132 members.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
+---
+layout: post
+title: Snowplow Scala Analytics SDK 0.4.0 released
+title-short: Snowplow Scala Analytics SDK 0.4.0
+tags: [scala, snowplow, enriched events, spark, dynamodb]
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: Nothing about DynamoDB here. I guess `spark` is also slightly confusing.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+2. [Using the typesafe API](#typesafe-api)
+3. [Additional changes](#additional-changes)
+4. [Upgrading](#upgrading)
+5. [Getting help](#help)
+
+<!--more-->
+
+<h2 id="event-api">1. Event API</h2>
+
+Previously, the JSON Event Transformer - an event transformer that takes a Snowplow enriched event and converts it into a JSON ready for further processing - used to return Strings, which represented enriched events turned into JSON objects. While this was a very unopinionated and minimalistic approach, it involved a lot of extra post-processing, namely:
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: > an event transformer that takes

"Transformer" is very vague technical term. I think something like "a module" or "a function" should fit better.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+- Accessing individual JSON fields required casting the result to an instance of the json4s AST class via unsafe operations, `parse(result)` or even `parse(result).asInstanceOf[JObject]`.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: >  or even `parse(result).asInstanceOf[JObject]`.
	
I'd remove this part.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+- Accessing always existing fields sill required redundant error processing logic, e.g. `parsedJson.map("event_id").getOrElse(throw new RuntimeException("event_id is not present in the enriched event")`
+- Getting a list of shredded types required having to use a separate function, `jsonifyWithInventory`.
+
+In 0.4.0, the `EventTransformer` API has been fully deprecated in favor of `Event` - a single typesafe container that contains all 132 members of a [canonical Snowplow event][canonical-event-model]. All fields are automatically converted to appropriate non-String types where possible - for instance, the `event_id` column is represented as a [UUID instance][java-uuid], while timestamps are converted into optional [Instant][java-instant] values, eliminating the need for common string conversions. Contexts and unstructured events are also wrapped in [self-describing data container types][self-describing-type], allowing for advanced operations such as Iglu URI lookups.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: > has been fully deprecated

A very diplomatic way to say that we replaced old API with new one.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+The case class has the following primary functions:
+
+- `Event.parse(line)` - similar to the old `transform` function, this method accepts an enriched Snowplow event as a string and returns an `Event` instance as a result.
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: > an enriched Snowplow event

an enriched Snowplow event in a canonical TSV+JSON format.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+<h2 id="typesafe-api">2. Using the typesafe API</h2>
+
+Since base results of the Scala Analytics SDK are now members of the `Event` case class, their output needs to be converted to JSON strings. For instance, the following code can be used in a data modelling job to load a series of events into a Spark dataframe:
+
+{% highlight scala %}
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+
+val events = input
+  .map(line => Event.parse(line))
+  .filter(_.isValid)
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: Can be simplified to:

```
val events = input
  .map(line => Event.parse(line))
  .flatMap(_.toOption)
  .map(event => event.toJson(true).noSpaces)
```

We don't need filter first - as we're throwing away this information anyway.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+  .flatMap(_.toOption)
+  .map(event => event.toJson(true).noSpaces)
+
+val dataframe = spark.read.json(events: _*)
```
Comment (chuwy) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: I don't this `read.json` can accept `Seq` (or something that can be unwrapped).

**Comment by Mandelliant** [comment]
@yalisassoon let me know what you think - if it's ready next week I'll schedule it in on Wednesday

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+5. [Getting help](#help)
+
+<!--more-->
+
+<h2 id="event-api">1. Event API</h2>
+
+Previously, the JSON Event Transformer - an event transformer that takes a Snowplow enriched event and converts it into a JSON ready for further processing - used to return Strings, which represented enriched events turned into JSON objects. While this was a very unopinionated and minimalistic approach, it involved a lot of extra post-processing, namely:
+
+- Accessing individual JSON fields required casting the result to an instance of the json4s AST class via unsafe operations, `parse(result)` or even `parse(result).asInstanceOf[JObject]`.
+- Accessing always existing fields sill required redundant error processing logic, e.g. `parsedJson.map("event_id").getOrElse(throw new RuntimeException("event_id is not present in the enriched event")`
```
Comment (goodits) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: Typo. Sill -> Still.

File: _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+- Getting a list of shredded types required having to use a separate function, `jsonifyWithInventory`.
```
Comment (goodits) [inline _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: This sentence doesn't make sense to me. Is this still correct?
> _To get_ a list of _the_ shredded types required _you need_ to use a separate function, `jsonifyWithInventory`.

Comment (rzats) [reply _posts/2019-02-09-snowplow-scala-analytics-sdk-0.4.0-released/index.md]: It's no longer correct. I've reworded it as

>To get a list of shredded types a separate function, `jsonifyWithInventory`, had to be used.

**Comment by Mandelliant** [comment]
Thank you to whoever renamed the post folder and updated the URL to today's date

**Pull request merged**